File: packages/python/src/slot_flight/slot/object/projections.py
from __future__ import annotations

from collections.abc import AsyncIterable
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class CompletedSlot:
    slot: str
    value: Any
    state: Any
    attempt: int


async def partial_object_iterator(events: AsyncIterable[dict[str, Any]]):
    async for event in events:
        if event_has_state(event):
            yield event["state"]


async def completed_slot_iterator(events: AsyncIterable[dict[str, Any]]):
    async for event in events:
        if event["type"] == "slot-complete":
            yield CompletedSlot(
                slot=event["slot"],
                value=event["value"],
                state=event["state"],
                attempt=event["attempt"],
            )


async def completed_event_iterator(events: AsyncIterable[dict[str, Any]]):
    async for event in events:
        if is_completed_output_event(event):
            yield event


def event_has_state(event: dict[str, Any]) -> bool:
    return event["type"] in {"slot-delta", "slot-complete", "done"}


def is_completed_output_event(event: dict[str, Any]) -> bool:
    return event["type"] in {"slot-complete", "slot-retry", "slot-error", "done"}

File: packages/typescript/src/slot/object/projections.ts
import type { SlotFlightEvent } from "../../types.js";
import type { CompletedSlot } from "./stream.js";

export async function* partialObjectIterator<T>(
  events: AsyncIterable<SlotFlightEvent>
): AsyncGenerator<Partial<T>> {
  for await (const event of events) {
    if (eventHasState(event)) {
      yield event.state as Partial<T>;
    }
  }
}

export async function* completedSlotIterator<T>(
  events: AsyncIterable<SlotFlightEvent>
): AsyncGenerator<CompletedSlot<T>> {
  for await (const event of events) {
    if (event.type === "slot-complete") {
      yield {
        slot: event.slot,
        value: event.value,
        state: event.state as Partial<T>
      };
    }
  }
}

export async function* completedEventIterator(
  events: AsyncIterable<SlotFlightEvent>
): AsyncGenerator<SlotFlightEvent> {
  for await (const event of events) {
    if (isCompletedOutputEvent(event)) {
      yield event;
    }
  }
}

function eventHasState(event: SlotFlightEvent): event is SlotFlightEvent & {
  state: unknown;
} {
  return (
    event.type === "slot-delta" ||
    event.type === "slot-complete" ||
    event.type === "done"
  );
}

function isCompletedOutputEvent(event: SlotFlightEvent): boolean {
  return (
    event.type === "slot-complete" ||
    event.type === "slot-retry" ||
    event.type === "slot-error" ||
    event.type === "done"
  );
}
